src/realm.js

```diff
@@ -583,35 +584,63 @@ export class Realm {
+        }
+      }
+      return false;
+    }
+
     let tvalFor: Map<any, AbstractValue> = new Map();
     pbindings.forEach((val, key, map) => {
       let value = val && val.value;
       if (value instanceof AbstractValue) {
         invariant(value._buildNode !== undefined);
-        let tval = gen.derive(value.types, value.values, [value], ([n]) => n, {
-          skipInvariant: true,
-        });
-        tval.mightBeEmpty = value.mightBeEmpty;
+        let tval = gen.derive(
+          value.types,
+          value.values,
+          [key.object, value],
+          ([o, n]) => {
+            invariant(value instanceof Value);
+            if (value.mightHaveBeenDeleted() && isSelfReferential(value, key.pathNode)) {
+              let inTest = t.binaryExpression("in", t.stringLiteral(key.key), o);
+              let addEmpty = t.conditionalExpression(inTest, n, emptyExpression);
+              n = t.logicalExpression("||", n, addEmpty);
+            }
+            return n;
+          },
+          {
+            skipInvariant: true,
+          }
+        );
         tvalFor.set(key, tval);
       }
     });
     pbindings.forEach((val, key, map) => {
       let path = key.pathNode;
       let tval = tvalFor.get(key);
       invariant(tval !== undefined);
-      let mightBeEmpty = tval.mightBeEmpty;
-      gen.emitStatement([key.object, tval], ([o, v]) => {
+      invariant(val !== undefined);
+      let value = val.value;
+      invariant(value instanceof Value);
+      let mightHaveBeenDeleted = value.mightHaveBeenDeleted();
+      let mightBeUndefined = value.mightBeUndefined();
+      gen.emitStatement([key.object, tval, this.intrinsics.empty], ([o, v, e]) => {
         invariant(path !== undefined);
         let lh = path.buildNode([o]);
         let r = t.expressionStatement(t.assignmentExpression("=", (lh: any), v));
-        if (mightBeEmpty) {
-          // If o does not have property key.key and v === undefined, omit the assignment (at runtime)
-          // if (v !== undefined || key.key in o) r
+        if (mightHaveBeenDeleted) {
           invariant(typeof key.key === "string"); // for now
-          let inTest = t.binaryExpression("in", t.stringLiteral(key.key), o);
-          let vDefined = t.binaryExpression("!==", v, t.unaryExpression("void", t.numericLiteral(0)));
-          let guard = t.logicalExpression("||", vDefined, inTest);
-          return t.ifStatement(guard, r);
+          // If v === __empty || (v === undefined  && !(key.key in o))  then delete it
+          let emptyTest = t.binaryExpression("===", v, e);
+          let undefinedTest = t.binaryExpression("===", v, voidExpression);
+          let inTest = t.unaryExpression("!", t.binaryExpression("in", t.stringLiteral(key.key), o));
+          let guard = t.logicalExpression("||", emptyTest, t.logicalExpression("&&", undefinedTest, inTest));
+          let deleteIt = t.expressionStatement(t.unaryExpression("delete", (lh: any)));
```

Comment:
Could you add an invariant that lh is the right type?

Reply:
Sadly no. The expected type is a Flow union and there is no way I know of to assert this dynamically.